File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/evaluators/MathMulContextEvaluator.java
package com.github.iamrenny.ruleflow.evaluators;

import com.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import com.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import com.github.iamrenny.ruleflow.visitors.Visitor;

import java.math.BigDecimal;
import java.math.RoundingMode;

public class MathMulContextEvaluator implements ContextEvaluator<RuleFlowLanguageParser.MathMulContext> {

    @Override
    public Object evaluate(RuleFlowLanguageParser.MathMulContext ctx, Visitor visitor) {
        Object leftVal = visitor.visit(ctx.left);
        Object rightVal = visitor.visit(ctx.right);

        Double left = Double.valueOf(leftVal.toString());
        Double right = Double.valueOf(rightVal.toString());

        switch (ctx.op.getType()) {
            case RuleFlowLanguageLexer.MULTIPLY:
                return left * right;
            case RuleFlowLanguageLexer.DIVIDE:
                if (right  == 0) {
                    return 0.00d;
                }
                return left / right;
            case RuleFlowLanguageLexer.MODULO:
                return left % right;
            default:
                throw new IllegalArgumentException("Operation not supported: " + ctx.op.getText());
        }
    }
}

File: analang-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/evaluators/MathMulContextEvaluator.kt
package com.rappi.fraud.rules.parser.evaluators

import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser
import com.rappi.fraud.rules.parser.visitors.Visitor
import java.math.BigDecimal
import java.math.RoundingMode

class MathMulContextEvaluator : ContextEvaluator<ANAParser.MathMulContext> {

    override fun evaluate(ctx: ANAParser.MathMulContext, visitor: Visitor): Any {
        val leftVal = visitor.visit(ctx.left)
        val rightVal = visitor.visit(ctx.right)

        val left = leftVal.toString().toBigDecimal().setScale(2, RoundingMode.DOWN)
        val right = rightVal.toString().toBigDecimal().setScale(2, RoundingMode.DOWN)

        return when {
            right.compareTo(BigDecimal.ZERO) == 0 -> return BigDecimal.ZERO
            ctx.op.type == ANALexer.MULTIPLY -> left.multiply(right)
            ctx.op.type ==    ANALexer.DIVIDE -> left.divide(right, 2, RoundingMode.DOWN)
            else -> throw IllegalArgumentException("Operation not supported: ${ctx.op.text}")
            }
    }

}